internal/controllers/configuration/kongupstreampolicy_controller.go

```diff
@@ -347,6 +460,27 @@ func (r *KongUpstreamPolicyReconciler) Reconcile(ctx context.Context, req ctrl.R
 		return ctrl.Result{}, nil
 	}
 
+	// Do not store the KongUpstreamPolicy into the cache if it is not used by some Service being the backend of Ingress/HTTPRoute
+	// that matches the ingress class/gateway class of the controller.
+	class := new(netv1.IngressClass)
+	if !r.DisableIngressClassLookups {
+		if err := r.Get(ctx, k8stypes.NamespacedName{Name: r.IngressClassName}, class); err != nil {
+			// we log this without taking action to support legacy configurations that only set ingressClassName or
+			// used the class annotation and did not create a corresponding IngressClass. We only need this to determine
+			// if the IngressClass is default or to configure default settings, and can assume no/no additional defaults
+			// if none exists.
+			log.V(logging.DebugLevel).Info("Could not retrieve IngressClass", "ingressclass", r.IngressClassName)
+		}
+	}
+	shouldStore, err := r.upstreamPolicyUsedByBackendsOfMatchingClass(ctx, req.NamespacedName, ctrlutils.IsDefaultIngressClass(class))
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	if !shouldStore {
+		log.V(logging.DebugLevel).Info("KongUpstreamPolicy is not referenced by Services used as backend of Ingress or HTTPRoute with matching class, skipping reconciliation")
+		return ctrl.Result{}, nil
+	}
```

Comment:
Can we add an envtest that covers this behavior (status is not populated if a parent HTTPRoute or Ingress is not to be reconciled by KIC)?

Reply:
Added an envtest case.